skxray/cdi.py

```diff
@@ -119,9 +125,339 @@ def convolution(array1, array2):
+    sup[sup_index] = 1
+    return sup
+
+
+def cdi_recon(diff_array, sample_obj, sup,
+              beta=1.15, start_ave=0.8, pi_modulus_flag='Complex',
+              sw_flag=True, sw_sigma=0.5, sw_threshold=0.1, sw_start=0.2,
+              sw_end=0.8, sw_step=10, n_iterations=1000):
+    """
+    Run reconstruction with difference map algorithm.
+
+    Parameters
+    ---------
+    diff_array : array
+        diffraction pattern from experiments
+    sample_obj : array
+        initial sample with phase
+    sup : array
+        initial support
+    beta : float, optional
+        feedback parameter for difference map algorithm.
+        default is 1.15.
+    start_ave : float, optional
+        define the point to start doing average.
+        default is 0.8.
+    pi_modulus_flag : str, optional
+        'Complex' or 'Real', defining the way to perform pi_modulus calculation.
+        default is 'Complex'.
+    sw_flag : Bool, optional
+        flag to use shrinkwrap algorithm or not.
+        default is True.
+    sw_sigma : float, optional
+        gaussian width used in sw algorithm.
+        default is 0.5.
+    sw_threshold : float, optional
+        shreshold cut in sw algorithm.
+        default is 0.1.
+    sw_start : float, optional
+        at which point to start to do shrinkwrap.
+        defualt is 0.2
+    sw_end : float, optional
+        at which point to stop shrinkwrap.
+        defualt is 0.8
+    sw_step : float, optional
+        the frequency to perform sw algorithm.
+        defualt is 10
+    n_iterations : int, optional
+        number of iterations to run.
+        default is 1000.
+
+    Returns
+    -------
+    obj_ave : array
+        reconstructed sample object
+    error_dict : dict
+        Error information for all iterations. The dict keys include
+        obj_error, diff_error and sup_error. Obj_error is a list of
+        the relative error of sample object. Diff_error is calculated as
+        the difference between new diffraction pattern and the original
+        diffraction pattern. And sup_error stores the size of the
+        sample support.
+    """
+
+    diff_array = np.array(diff_array)     # diffraction data
+
+    gamma_1 = -1/beta
+    gamma_2 = 1/beta
+
+    # get support index
+    sup_index = np.where(sup == 1)
+    sup_out_index = np.where(sup != 1)
+
+    error_dict = {}
+    obj_error = np.zeros(n_iterations)
+    diff_error = np.zeros(n_iterations)
+    sup_error = np.zeros(n_iterations)
+
+    sup_old = np.zeros_like(diff_array)
+    obj_ave = np.zeros_like(diff_array).astype(complex)
+    ave_i = 0
+
+    time_start = time.time()
+    for n in range(n_iterations):
+        obj_old = np.array(sample_obj)
+
+        obj_a = pi_modulus(sample_obj, diff_array)
+        if pi_modulus_flag.lower() == 'real':
+            obj_a = np.abs(obj_a)
+
+        obj_a = (1 + gamma_2) * obj_a - gamma_2 * sample_obj
+        obj_a = pi_support(obj_a, sup_out_index)
```

Comment:
Indexing is faster than multiply operation. In cdi case, indexing is around 5 times faster, and even better when it comes to 3D case.